### ISSUE TITLE
express middleware doesn't allow user to configure custom transports

### ISSUE THREAD

**Issue by ofrobots** [googleapis/nodejs-logging-winston#303]
Originally reported here: https://github.com/googleapis/nodejs-logging-winston/pull/182#issuecomment-484692277

* Either we should provide users to configure a set of transports for the winston loggers we create.
* Or we should not be in the business of creating winston loggers. The users should instantiate the logger and pass to us. This is probably too complicated because we conditionally need to create two loggers.

@googleapis/node-team opinions?

**Comment by daaain** [comment]
I'd definitely prefer if there was a way to pass in a root logger which the middleware one would inherit from!

But FWIW I found a workaround to the transport issue which does the job:

```js
// index.ts
app.use('/graphql', (req, res, next) => {
  // this is where we call the Google logging middleware so that it can generate the tracing code
  if (googleMiddleware) {
    return googleMiddleware(req, res, next)
  }
  next()
})

app.use('/graphql', (req, _, next) => {
  if (googleMiddleware) {
    // in development we need to add the Console transport to Google logging middleware's Winston
    // instance injected above so we can see the output
    if (DEV) {
      req.log.add(new winston.transports.Console())
    }
  }
  next()
})
```

For further context I also define a global `logger` which then gets updated using the instance returned by the middleware. I'm using Winston's `configure` method to write my settings back, which then also get inherited by the child instances created per request:

```js
// logger.ts
import util from 'util'

import winston from 'winston'
import { LEVEL, SPLAT } from 'triple-beam'

export const DEV = process.env.NODE_ENV !== 'production'

export const defaultMeta = {
  app_name: process.env.APP_NAME,
  release: process.env.RELEASE,
}

let winstonFormat = winston.format.json()
if (DEV) {
  winstonFormat = winston.format.combine(
    winston.format.colorize(),
    winston.format.timestamp(),
    winston.format.printf(
      ({
        level,
        message,
        timestamp,
        // we're destructuring these Symbols so that they don't show up in meta
        [LEVEL]: levelSymbol,
        [SPLAT]: splatSymbol,
        ...meta
      }) => {
        return `${timestamp} ${level}: ${message}\n${util.inspect(meta, {
          colors: true,
          depth: 5,
        })}`
      },
    ),
  )
}

export let logger
export function createLogger(loggerInstance) {
  logger = loggerInstance
  logger.configure({
    level: process.env.DEBUG ? 'debug' : 'info',
    transports: loggerInstance.transports,
    format: winstonFormat,
    defaultMeta,
  })
}

createLogger(
  winston.createLogger({
    transports: [new winston.transports.Console()],
  }),
)

export default logger
```

To make this work without the very awkward async init function I bootstrap the middleware like this:

```js
// index.ts
let googleMiddleware
LoggingWinston.express
  .middleware({
    serviceContext: {
      service: process.env.APP_NAME,
      version: process.env.RELEASE,
    },
    level: process.env.DEBUG ? 'debug' : 'info',
  })
  .then(({ logger: googleLogger, mw }) => {
    // we need to update our logger with the instance supplied by the Google library
    createLogger(googleLogger)
    googleMiddleware = mw
    logger.info('Google logger Express middleware initialised')
  })
  .catch(error => {
    logger.warn(
      'Could not initialise Google logger Express middleware, error:',
      { error },
    )
  })
```

I'd highly recommend updating the code snippet in the README as the async function creates a lot of issues, amongst others breaking Nodemon, making Express non-responsive for ~2 seconds longer, etc.

**Comment by daaain** [comment]
What I realised while testing in GKE is that `winston.transports.Console` needs to be added in the deployed service too, otherwise the output won't be under container logs (only `projects/<project id>/logs/winston_log_applog`), unless I'm missing something?

**Comment by daaain** [comment]
The container logs via `winston.transports.Console` will not have the trace metadata though 😿 

Is there any chance setting the `resource` bit?

It should look something like this in a container log entry:

```
 resource: {
  labels: {
   cluster_name: "master"    
   container_name: "graphql"    
   location: "europe-west2-a"    
   namespace_name: "default"    
   pod_name: "graphql-6f94888ffc-crqh5"    
   project_id: "<project id>"    
  }
  type: "k8s_container"   
 }
```

But it looks like this in the custom log instead:

```
 resource: {
  labels: {
   cluster_name: "master"    
   container_name: ""    
   instance_id: ""    
   namespace_id: "default"    
   pod_id: ""    
   project_id: "<project id>"    
   zone: ""    
  }
  type: "container"   
 }
```

**Comment by EdByrnee** [comment]
There needs to be a better way of doing this, the async function is too awkward and you can't instantiate your logger with your custom transports etc.

There needs to be a wrapper for the winston.createLogger function that returns your own logger, but configured for express.

**Comment by EdByrnee** [comment]
> What I realised while testing in GKE is that `winston.transports.Console` needs to be added in the deployed service too, otherwise the output won't be under container logs (only `projects/<project id>/logs/winston_log_applog`), unless I'm missing something?

Hi daaain did you get any further with this?

**Comment by daaain** [comment]
@EdByrnee yeah, with the latest version (1.1.0) it's *almost* there...

This is how my code looks now:

```js
// logger.ts
import util from 'util'

import winston from 'winston'
import { LEVEL, SPLAT } from 'triple-beam'

export const DEV = process.env.NODE_ENV !== 'production'

export const defaultMeta = {
  app_name: process.env.APP_NAME,
  release: process.env.RELEASE,
}

let winstonFormat = winston.format.json()
if (DEV) {
  winstonFormat = winston.format.combine(
    winston.format.colorize(),
    winston.format.timestamp(),
    winston.format.printf(
      ({
        level,
        message,
        timestamp,
        // we're destructuring these Symbols so that they don't show up in meta
        [LEVEL]: levelSymbol,
        [SPLAT]: splatSymbol,
        ...meta
      }) => {
        return `${timestamp} ${level}: ${message}\n${util.inspect(meta, {
          colors: true,
          depth: 5,
        })}`
      },
    ),
  )
}

export let logger
export function createLogger(loggerInstance) {
  logger = loggerInstance
  logger.configure({
    level: process.env.DEBUG ? 'debug' : 'info',
    transports: [
      new winston.transports.Console(),
      ...loggerInstance.transports,
    ],
    format: winstonFormat,
    defaultMeta,
  })
}

createLogger(winston.createLogger({}))

export default logger
```

```js
// index.ts
...

let googleMiddleware
LoggingWinston.express
  .middleware({
    serviceContext: {
      service: process.env.APP_NAME,
      version: process.env.RELEASE,
    },
    level: process.env.DEBUG ? 'debug' : 'info',
  })
  .then(({ logger: googleLogger, mw }) => {
    // we need to update our logger with the instance supplied by the Google library
    createLogger(googleLogger)
    googleMiddleware = mw
    logger.info('Google logger Express middleware initialised')
  })
  .catch(error => {
    logger.warn(
      'Could not initialise Google logger Express middleware, error:',
      { error },
    )
  })

app.use('/graphql', (req, res, next) => {
  // this is where we call the Google logging middleware so that it can generate the tracing code
  if (googleMiddleware) {
    return googleMiddleware(req, res, next)
  }
  next()
})

// OPTIONAL
app.use('/graphql', (req, res, next) => {
  // insert tracing header into response so we can log on the client
  if (googleMiddleware) {
    res.set('x-cloud-trace-context', get(req, 'headers.x-cloud-trace-context'))
  }
  next()
})

...

```

In the Global view I get the requests bundled now (although in the wrong order 😞 ), but in the container log (via `winston.transports.Console`) it's just a linear log. Which is maybe because the middleware only sends the special bundler log message via the built in transport?

<img width="1386" alt="Screenshot 2019-06-24 at 14 34 49" src="https://user-images.githubusercontent.com/69962/60023454-bdc6fa80-968d-11e9-9d6d-e8d8f6d4fdab.png">

**Comment by daaain** [comment]
Oh, and almost forgot to mention that the containers do not log in the Global logs (only the local development instance in Docker Compose), but in the GKE Container one. Which annoyingly is not the same as the one you can access when clicking on the Container logs link under a GKE Workload. So currently I have logs from the same application in 3 different places / logs.

I think that's because the `resource` metadata isn't filled in properly.

This is how it looks like in log entries under GKE Container stored by the middleware:

```
 resource: {
  labels: {
   cluster_name: "master"    
   container_name: ""    
   instance_id: ""    
   namespace_id: "default"    
   pod_id: ""    
   project_id: "basis-janssen-test"    
   zone: ""    
  }
  type: "container"   
 }
```

And this is how it looks like in the "proper" container logs, currently stored via Console transport:

```
 resource: {
  labels: {
   cluster_name: "master"    
   container_name: "graphql"    
   location: "europe-west2-a"    
   namespace_name: "default"    
   pod_name: "graphql-85bf76c55c-vqstq"    
   project_id: "<project id>"    
  }
  type: "k8s_container"   
 }
```

Maybe I should open this as a new issue? Though this might be an upstream one?

**Comment by ofrobots** [comment]
> the async function is too awkward

Async function is necessary because, various things can fail (asynchronously) trying to instantiate a connection to Stackdriver Logging service. The middleware needs to open this connection on your behalf. If we do not make the construction async, things would be even more awkward. Failure to initialize logging would be considered to be a fatal problem by many folks.

If the construction is sync, we have only two choices:

1. Throw error asynchronously. You cannot catch async errors except through global uncaught exception handlers.
1. Raise the error when you attempt to log. Most people don't install callbacks on individual log messages, so this error would go silently ignored by most. Even if that was not an issue, it would quite awkward to have each log statement in your app installing a calling to deal with an async init failure.

The only alternative to async init I can think of is for the users to initialize a `@google-cloud/logging` instance and pass that in. That way, the middleware would not be responsible for async init, and the onus of dealing with errors would be entirely on the users.

**Comment by daaain** [comment]
@ofrobots I completely understand the library's internal init being async, my beef was using the async function in user code as in the code example in the readme.

As you can see in the code examples I posted above it's completely possible to not have an async function, but work with the promise. But ideally this should be something the library does for me.

The main issue with doing it this way is that the `logger` instance is undefined until the async middleware init returns its own, so that's why I create a Console only Winston instance and replace it with the middleware's one once it's ready.

Not only does this solve the async problem, but makes the logger importable anywhere in the source code without having to wait for the async function to return the instance everywhere in the codebase which would be horrendous.

**Comment by ofrobots** [comment]
In order to keep the discussion focussed, I've split the async init issue to #355. Let's keep this issue focused on allowign users to customize transports.

@daaain 
> I'd definitely prefer if there was a way to pass in a root logger which the middleware one would inherit from!

We actually need (conditionally) two loggers to be provided as we need to use different transports for the application log and the request log. This is a requirement for request bundling to work correctly. I am afraid this would be too complex for users to understand.

An alternative would be for a `createLogger` function to be provided by the user that we could call. You could install your transports there. Would that work ?

**Comment by daaain** [comment]
@ofrobots couldn't both of these loggers be created as children of the root logger passed in, like you do with the request loggers?

Or, since the `LoggingWinston.express.middleware` constructor accepts an `options` object parameter, there could be a `config` property on it which can be used to pass any standard Winston config. This would be used as the starting config which the middleware can then enhance with additional transports, metadata, etc and create children of for each request.

Ideally the logger instance should be returned immediately alongside the async promise to be available for use (and export) right away, instead of when the promise returns. The log messages could be queued up until async the Stackdriver transport is ready, but emitted immediately on other transports.

This is pretty much what I implemented in the code examples above (except the queue), but would be great if the library took care of it and I only needed to pass my formatter and conditionally add console transport if it's in dev.
